CPU Limitation

Tell me if it is possible to limit the CPU usage in the docker compass, so that there are few resources left for other applications.
I found an opportunity to add a record to the service

```
version: '3'
services:
  myservice:
    image: myimage
    cpus: 0.8
```

In this example, we limit the myservice container's usage to no more than 80% of the host machine's CPU time using the cpus option. The value 0.8 means that the container will have access to 80% of the processor's core time.

or

```
version: '3'
services:
  myservice:
    image: myimage
    cpuset: "0-2"
```

In this example, the myservice container will be limited to using CPU cores 0 to 2.

cpuset is a parameter that allows you to explicitly set on which CPU cores or CPU core groups a container will run. This is useful when you want to control which processor cores are used for different containers.


But with such a limitation, the backend crashes with an error 

```
backend         | [2023-09-18 13:22:47 +0000] [113] [INFO] Worker exiting (pid: 113)
backend         | [2023-09-18 13:22:48 +0000] [80] [ERROR] Worker (pid:113) was sent SIGKILL! Perhaps out of memory?

```

Docker Compose also supports other options for limiting CPU time:

1. `cpu_shares`: This option allows you to set the relative priority of CPU time usage for containers. The higher the `cpu_shares` value, the higher priority the container gets in CPU time allocation. By default, all containers have the same `cpu_shares` value. Example:

   ```
   version: '3'
   services:
     myservice:
       image: myimage
       cpu_shares: 512
   ```

2. `cpu_quota` and `cpu_period`: With these options, you can explicitly set a quota (amount of CPU time) and period (duration of time period) for the container. The quota is specified in microseconds of CPU time per period, which is also specified in microseconds. Example:

   ```
   version: '3'
   services:
     myservice:
       image: myimage
       cpu_quota: 50000
       cpu_period: 100000
   ```

   In this example, the `myservice` container is allowed to use no more than half of the host machine's CPU time.

Note that using these options in Docker Compose requires support from your host system and Docker Compose version 3.2 or above. Consult the Docker documentation for more detailed information and clarifications regarding the availability of these options in your Docker runtime environment.

The last two options do not seem to have given a visual result.

Maybe someone was interested in this question

--- a/apps/backend/librephotos/settings/production.py
+++ b/apps/backend/librephotos/settings/production.py
@@ -91,6 +91,14 @@ Q_CLUSTER = {
     "poll": 1,
 }
 
+# Number of background workers doing the heavy lifting (thumbnails, face
+# detection, captioning). Left unset, django-q falls back to cpu_count(), which
+# reports the *host's* cores even when the container is capped with a compose
+# `cpus:` limit - so the pool stays just as wide and the limit only starves it.
+# Setting this is the way to actually reduce LibrePhotos' CPU and RAM appetite.
+if os.environ.get("WORKER_CONCURRENCY"):
+    Q_CLUSTER["workers"] = int(os.environ["WORKER_CONCURRENCY"])
+
 CONSTANCE_BACKEND = "constance.backends.database.DatabaseBackend"
 CONSTANCE_ADDITIONAL_FIELDS = {
     "map_api_provider": [

--- a/apps/docs/docs/installation/environment-variables.md
+++ b/apps/docs/docs/installation/environment-variables.md
@@ -56,6 +56,82 @@ To leverage GPU acceleration for neural networks and face detection, follow thes
 The GPU image is only available for x86 architecture. ARM is not supported for the GPU image.
 :::
 
+### Limiting CPU and memory usage
+
+The backend container runs two things: **gunicorn**, which answers API requests, and a pool of **background workers**, which scan your library — thumbnails, face detection, captioning. Almost all of the CPU and memory LibrePhotos uses goes to the background workers, and by default there is **one worker per CPU core**.
+
+#### Start with the worker count, not a CPU limit
+
+The instinct is to cap the container in `docker-compose.yml`:
+
+```yaml
+services:
+  backend:
+    cpus: 0.8
+```
+
+On its own this usually backfires. A `cpus:` limit throttles the container, but it does not change how many workers LibrePhotos starts — the worker pool is sized from the number of cores the *host* reports, which a `cpus:` limit does not change. You end up with just as many workers competing for a fraction of the CPU, and the first thing to break is the API: a request that takes longer than gunicorn's timeout gets its worker killed, and the backend log fills with
+
+```
+[ERROR] Worker (pid:113) was sent SIGKILL! Perhaps out of memory?
+```
+
+That message is gunicorn's generic text for a killed worker. On a CPU-capped host it almost always means the request was too slow, **not** that the machine ran out of memory.
+
+So set the worker count instead. In your `.env`:
+
+```bash
+# One background worker instead of one per core
+workerConcurrency=1
+```
+
+This is the setting that actually gives resources back to the rest of the machine. Scanning takes longer, but the API stays responsive and nothing gets killed. Each worker is recycled once it passes roughly 300 MB of resident memory, so the worker count is also the main lever on the backend's memory footprint.
+
+#### If you also want a hard cap
+
+Once the worker count is sensible, a container limit is a reasonable backstop. Raise the API timeout at the same time so throttled requests are not killed mid-flight:
+
+```bash
+workerConcurrency=1
+# Seconds before gunicorn kills a request (default 30)
+gunicornTimeout=120
+```
+
+```yaml
+services:
+  backend:
+    cpus: 2.0
+    mem_limit: 4g
+```
+
+`cpuset` works too, if you would rather pin LibrePhotos to specific cores than give it a fraction of all of them:
+
+```yaml
+services:
+  backend:
+    cpuset: "0-2"
+```
+
+Under Docker Swarm the equivalent is a `deploy.resources` block:
+
+```yaml
+services:
+  backend:
+    deploy:
+      resources:
+        limits:
+          cpus: "2.0"
+          memory: 4G
+```
+
+:::note
+`cpu_shares` only sets a *relative* priority between containers and has no visible effect when nothing else is competing for the CPU, which is why it often looks like it does nothing.
+:::
+
+:::warning
+Do not cap the container so hard that the first scan cannot finish. Face detection and captioning load sizeable models; below roughly 2 GB of memory the backend will be killed by the kernel — and *that* really is an out-of-memory kill.
+:::
+
 ### Hosting under a sub-path (subdirectory)
 
 By default LibrePhotos is served from the root of a domain (e.g. `https://photos.example.com/`). If you need to host it under a sub-path instead (e.g. `https://example.com/photos/`), the frontend has to know that base path.

--- a/apps/docs/docs/user-guide/settings/date-rules.md
+++ b/apps/docs/docs/user-guide/settings/date-rules.md
@@ -38,6 +38,16 @@ should be one of the following:
 
 If either "source_tz" or "report_tz" could not be obtained, the rule is considered to be not applicable.
 
+### What gets stored, and why times are not converted for you
+
+Once a rule succeeds, LibrePhotos stores the **local time the camera recorded** — the time you would have read off the camera when you took the shot.
+
+This is deliberate. A photo taken at 14:00 in Tokyo and one taken at 14:00 in Berlin are both shown as 14:00, because that is the time each one was actually taken. LibrePhotos does **not** re-cast your library into the timezone you happen to be browsing from; if it did, every photo would shift whenever you travelled, and photos from a holiday abroad would no longer line up with the day you remember them on.
+
+:::note
+Under the hood the local time is tagged as UTC before it is saved. That is an internal storage detail, not a claim that the photo was taken in UTC — it is what keeps the wall-clock time stable for every viewer. If you query the database directly, read the timestamps as-is and do not convert them.
+:::
+
 ### The different rule types
 
 Currently, supported rule types:
@@ -60,4 +70,16 @@ The top rules will be applied first, and the bottom rules be applied last. To ch
 
 ### Applying the rules
 
-The date rules are applied on each scan. If you changed the rules to your liking, click on "Rescan all Photos" and then the new rules will be applied
+The date rules are applied on each scan. If you changed the rules to your liking, click on "Rescan all Photos" and then the new rules will be applied.
+
+Use **"Rescan all Photos"** specifically. An ordinary scan only looks at files that are new or whose contents changed since the last scan, so it will walk straight past a library that is already indexed and nothing will appear to happen. "Rescan all Photos" re-reads every file and re-runs the rules over all of them.
+
+### Troubleshooting
+
+**Every photo is off by the same number of hours.**
+
+If your whole library is shifted by a constant amount — and that amount happens to match your own UTC offset — no date rule will fix it, because the rules are not what is wrong. This was a display bug: the times were stored correctly, but the web interface converted them into the timezone of the browser you were viewing from, so a viewer at UTC+3 saw every photo three hours late. It is fixed in current versions; update and reload the page, and the stored times will show as they always were. No rescan is needed.
+
+**A rule change had no effect at all.**
+
+Check that you used "Rescan all Photos" rather than a normal scan (see above), and that the rule you expect to win is above the one currently matching — the first rule that can produce a time is the one that is used. A rule that names an EXIF tag your files do not carry is silently skipped, so it is worth confirming the tag actually exists in your photos.

--- a/apps/frontend/src/components/lightbox/TimestampItem.tsx
+++ b/apps/frontend/src/components/lightbox/TimestampItem.tsx
@@ -18,6 +18,7 @@ import "react-virtualized/styles.css";
 import { useUpdatePhotoMutation } from "../../api_client/photos/hooks";
 import { Photo } from "../../api_client/photos/types";
 import { i18nResolvedLanguage } from "../../i18n";
+import { parsePhotoTimestamp, photoTimestampToPickerDate, pickerDateToPhotoTimestamp } from "../../util/dateUtils";
 
 const isValidDate = (date: Date | null): date is Date => date instanceof Date && !Number.isNaN(date.getTime());
 
@@ -27,13 +28,9 @@ type Props = Readonly<{
 }>;
 
 export function TimestampItem({ photoDetail, isPublic }: Props) {
-  const [timestamp, setTimestamp] = useState(() => {
-    if (!photoDetail.exif_timestamp) {
-      return null;
-    }
-    const date = new Date(photoDetail.exif_timestamp);
-    return Number.isNaN(date.getTime()) ? null : date;
-  });
+  const [timestamp, setTimestamp] = useState(() =>
+    photoDetail.exif_timestamp ? photoTimestampToPickerDate(photoDetail.exif_timestamp) : null
+  );
 
   // savedTimestamp is used to cancel timestamp modification
   const [savedTimestamp, setSavedTimestamp] = useState(timestamp);
@@ -73,7 +70,7 @@ export function TimestampItem({ photoDetail, isPublic }: Props) {
 
   const onSaveDateTime = () => {
     const differentJson = {
-      exif_timestamp: isValidDate(timestamp) ? timestamp.toISOString() : null,
+      exif_timestamp: isValidDate(timestamp) ? pickerDateToPhotoTimestamp(timestamp) : null,
     };
     updatePhoto({ id: photoDetail.image_hash!, data: differentJson });
     setEditMode(false);
@@ -88,11 +85,11 @@ export function TimestampItem({ photoDetail, isPublic }: Props) {
   const getDateTimeLabel = () => {
     if (!photoDetail.exif_timestamp) return t("lightbox.sidebar.withouttimestamp");
 
-    const photoDateTime = DateTime.fromISO(photoDetail.exif_timestamp);
+    const photoDateTime = parsePhotoTimestamp(photoDetail.exif_timestamp).setLocale(lang);
     if (photoDateTime.isValid) {
-      const date = DateTime.fromISO(photoDetail.exif_timestamp).setLocale(lang).toLocaleString(DateTime.DATE_MED);
-      const dayOfWeek = DateTime.fromISO(photoDetail.exif_timestamp).setLocale(lang).toFormat("cccc");
-      const time = DateTime.fromISO(photoDetail.exif_timestamp).setLocale(lang).toLocaleString(DateTime.TIME_SIMPLE);
+      const date = photoDateTime.toLocaleString(DateTime.DATE_MED);
+      const dayOfWeek = photoDateTime.toFormat("cccc");
+      const time = photoDateTime.toLocaleString(DateTime.TIME_SIMPLE);
       return (
         <div>
           {date}{" "}
@@ -113,7 +110,7 @@ export function TimestampItem({ photoDetail, isPublic }: Props) {
 
   const onUndoChangedTimestamp = () => {
     const differentJson = {
-      exif_timestamp: isValidDate(savedTimestamp) ? savedTimestamp.toISOString() : null,
+      exif_timestamp: isValidDate(savedTimestamp) ? pickerDateToPhotoTimestamp(savedTimestamp) : null,
     };
     updatePhoto({ id: photoDetail.image_hash!, data: differentJson });
     setTimestamp(savedTimestamp);

--- a/apps/frontend/src/components/lightbox/TimestampItem.tsx
+++ b/apps/frontend/src/components/lightbox/TimestampItem.tsx
@@ -18,7 +18,12 @@ import "react-virtualized/styles.css";
 import { useUpdatePhotoMutation } from "../../api_client/photos/hooks";
 import { Photo } from "../../api_client/photos/types";
 import { i18nResolvedLanguage } from "../../i18n";
-import { parsePhotoTimestamp, photoTimestampToPickerDate, pickerDateToPhotoTimestamp } from "../../util/dateUtils";
+import {
+  parsePhotoTimestamp,
+  parsePickerDate,
+  photoTimestampToPickerDate,
+  pickerDateToPhotoTimestamp,
+} from "../../util/dateUtils";
 
 const isValidDate = (date: Date | null): date is Date => date instanceof Date && !Number.isNaN(date.getTime());
 
@@ -48,13 +53,17 @@ export function TimestampItem({ photoDetail, isPublic }: Props) {
       return;
     }
 
-    const newDate = new Date(date);
+    const newDate = parsePickerDate(date);
+    if (!newDate) {
+      setTimestamp(null);
+      return;
+    }
     if (timestamp) {
       newDate.setHours(timestamp.getHours());
       newDate.setMinutes(timestamp.getMinutes());
       newDate.setSeconds(timestamp.getSeconds());
     }
-    setTimestamp(Number.isNaN(newDate.getTime()) ? null : newDate);
+    setTimestamp(newDate);
   };
 
   const onChangeTime = (event: React.ChangeEvent<HTMLInputElement>) => {

--- a/apps/frontend/src/components/stacks-duplicates/DuplicatesPageContent.tsx
+++ b/apps/frontend/src/components/stacks-duplicates/DuplicatesPageContent.tsx
@@ -41,6 +41,7 @@ import {
   IconTrash,
   IconX,
 } from "@tabler/icons-react";
+import { DateTime } from "luxon";
 import React, { useEffect, useState } from "react";
 import { useTranslation } from "react-i18next";
 import { serverAddress } from "../../api_client/apiClient";
@@ -58,6 +59,7 @@ import {
 import type { Duplicate, DuplicatePhoto, DuplicateType, ReviewStatus } from "../../api_client/duplicates/types";
 import { duplicateTypeLabels } from "../../api_client/duplicates/types";
 import { useFetchUserSelfDetailsQuery } from "../../api_client/user/hooks";
+import { parsePhotoTimestamp } from "../../util/dateUtils";
 import { Lightbox } from "../lightbox";
 
 function formatFileSize(bytes: number): string {
@@ -190,7 +192,7 @@ function DuplicatePhotoCard({
 
         {photo.exif_timestamp && (
           <Text size="xs" c="dimmed">
-            📅 {new Date(photo.exif_timestamp).toLocaleDateString()}
+            📅 {parsePhotoTimestamp(photo.exif_timestamp).toLocaleString(DateTime.DATE_SHORT)}
           </Text>
         )}
 

--- a/apps/frontend/src/components/stacks/StackModal.tsx
+++ b/apps/frontend/src/components/stacks/StackModal.tsx
@@ -29,11 +29,13 @@ import {
   IconSun,
   IconVideo,
 } from "@tabler/icons-react";
+import { DateTime } from "luxon";
 import React, { useState } from "react";
 import { useTranslation } from "react-i18next";
 import { serverAddress } from "../../api_client/apiClient";
 import { useSetCoverPhotoMutation, useStackQuery } from "../../api_client/stacks";
 import type { StackType } from "../../api_client/stacks/types";
+import { parsePhotoTimestamp } from "../../util/dateUtils";
 import { StackLightbox } from "./StackLightbox";
 
 // File variant type
@@ -231,7 +233,7 @@ function StackPhotoCard({
 
         {photo.exif_timestamp && (
           <Text size="xs" c="dimmed">
-            📅 {new Date(photo.exif_timestamp).toLocaleDateString()}
+            📅 {parsePhotoTimestamp(photo.exif_timestamp).toLocaleString(DateTime.DATE_SHORT)}
           </Text>
         )}
 

--- a/apps/frontend/src/routes/_protected/photo.$id.tsx
+++ b/apps/frontend/src/routes/_protected/photo.$id.tsx
@@ -30,6 +30,7 @@ import { LocationSection } from "../../components/lightbox/LocationSection";
 import { PeopleSection } from "../../components/lightbox/PeopleSection";
 import { SimilarPhotosSection } from "../../components/lightbox/SimilarPhotosSection";
 import { TimestampItem } from "../../components/lightbox/TimestampItem";
+import { parsePhotoTimestamp } from "../../util/dateUtils";
 
 export const Route = createFileRoute("/_protected/photo/$id")();
 
@@ -68,7 +69,7 @@ export function SinglePhotoView() {
       ? photoDetail.image_path[0].substring(photoDetail.image_path[0].lastIndexOf("/") + 1)
       : "Unknown filename";
   const timestamp = photoDetail.exif_timestamp
-    ? DateTime.fromISO(photoDetail.exif_timestamp).toLocaleString(DateTime.DATETIME_MED)
+    ? parsePhotoTimestamp(photoDetail.exif_timestamp).toLocaleString(DateTime.DATETIME_MED)
     : "Unknown timestamp";
 
   return (

--- a/apps/frontend/src/util/dateUtils.test.ts
+++ b/apps/frontend/src/util/dateUtils.test.ts
@@ -1,0 +1,71 @@
+import { Settings } from "luxon";
+import { afterEach, describe, expect, test } from "vitest";
+import { parsePhotoTimestamp, photoTimestampToPickerDate, pickerDateToPhotoTimestamp } from "./dateUtils";
+
+const originalZone = Settings.defaultZone;
+
+afterEach(() => {
+  Settings.defaultZone = originalZone;
+});
+
+// The bug in #1025: a viewer in UTC+3 saw every photo shifted three hours
+// forward, and no date rule change could correct it. These zones bracket that
+// case on both sides of UTC.
+const zones = ["UTC", "Europe/Moscow", "America/New_York", "Asia/Kolkata"];
+
+describe("parsePhotoTimestamp", () => {
+  test("returns the recorded wall clock whatever zone the viewer is in", () => {
+    zones.forEach(zone => {
+      Settings.defaultZone = zone;
+      const parsed = parsePhotoTimestamp("2023-09-21T14:30:00Z");
+      expect(parsed.isValid).toBe(true);
+      expect([parsed.year, parsed.month, parsed.day]).toEqual([2023, 9, 21]);
+      expect([parsed.hour, parsed.minute]).toEqual([14, 30]);
+    });
+  });
+
+  test("does not roll the date over near midnight", () => {
+    Settings.defaultZone = "Europe/Moscow";
+    const parsed = parsePhotoTimestamp("2023-09-21T23:30:00Z");
+    expect(parsed.day).toBe(21);
+    expect(parsed.hour).toBe(23);
+  });
+
+  test("flags an unparseable timestamp as invalid", () => {
+    expect(parsePhotoTimestamp("not a timestamp").isValid).toBe(false);
+  });
+});
+
+describe("photoTimestampToPickerDate", () => {
+  test("exposes the wall clock through the Date's local getters", () => {
+    const date = photoTimestampToPickerDate("2023-09-21T14:30:15Z");
+    expect(date).not.toBeNull();
+    expect(date!.getFullYear()).toBe(2023);
+    expect(date!.getMonth()).toBe(8);
+    expect(date!.getDate()).toBe(21);
+    expect(date!.getHours()).toBe(14);
+    expect(date!.getMinutes()).toBe(30);
+    expect(date!.getSeconds()).toBe(15);
+  });
+
+  test("returns null for an unparseable timestamp", () => {
+    expect(photoTimestampToPickerDate("nope")).toBeNull();
+  });
+});
+
+describe("pickerDateToPhotoTimestamp", () => {
+  test("round-trips a timestamp unchanged", () => {
+    const timestamp = "2023-09-21T14:30:15.000Z";
+    expect(pickerDateToPhotoTimestamp(photoTimestampToPickerDate(timestamp)!)).toBe(timestamp);
+  });
+
+  test("stores the wall clock the user picked, not a converted instant", () => {
+    // The user picks 09:00 in the time input; that is what should be stored.
+    const picked = new Date(2023, 8, 21, 9, 0, 0);
+    expect(pickerDateToPhotoTimestamp(picked)).toBe("2023-09-21T09:00:00.000Z");
+  });
+
+  test("returns null for an invalid Date", () => {
+    expect(pickerDateToPhotoTimestamp(new Date(Number.NaN))).toBeNull();
+  });
+});

--- a/apps/frontend/src/util/dateUtils.test.ts
+++ b/apps/frontend/src/util/dateUtils.test.ts
@@ -1,6 +1,11 @@
 import { Settings } from "luxon";
 import { afterEach, describe, expect, test } from "vitest";
-import { parsePhotoTimestamp, photoTimestampToPickerDate, pickerDateToPhotoTimestamp } from "./dateUtils";
+import {
+  parsePhotoTimestamp,
+  parsePickerDate,
+  photoTimestampToPickerDate,
+  pickerDateToPhotoTimestamp,
+} from "./dateUtils";
 
 const originalZone = Settings.defaultZone;
 
@@ -50,6 +55,28 @@ describe("photoTimestampToPickerDate", () => {
 
   test("returns null for an unparseable timestamp", () => {
     expect(photoTimestampToPickerDate("nope")).toBeNull();
+  });
+});
+
+describe("parsePickerDate", () => {
+  test("keeps the calendar day the user clicked", () => {
+    // `new Date("2023-09-21")` resolves to UTC midnight, which is still the
+    // 20th anywhere west of UTC.
+    const parsed = parsePickerDate("2023-09-21");
+    expect(parsed).not.toBeNull();
+    expect(parsed!.getFullYear()).toBe(2023);
+    expect(parsed!.getMonth()).toBe(8);
+    expect(parsed!.getDate()).toBe(21);
+  });
+
+  test("passes a Date through unchanged", () => {
+    const date = new Date(2023, 8, 21, 14, 30);
+    expect(parsePickerDate(date)!.getTime()).toBe(date.getTime());
+  });
+
+  test("returns null for unusable input", () => {
+    expect(parsePickerDate("not a date")).toBeNull();
+    expect(parsePickerDate(new Date(Number.NaN))).toBeNull();
   });
 });
 

--- a/apps/frontend/src/util/dateUtils.ts
+++ b/apps/frontend/src/util/dateUtils.ts
@@ -45,6 +45,19 @@ export function photoTimestampToPickerDate(timestamp: string): Date | null {
 }
 
 /**
+ * Read a value out of a date picker as a calendar day.
+ *
+ * Mantine hands back either a `Date` or a date-only string like `"2023-09-21"`,
+ * and `new Date("2023-09-21")` resolves that string to UTC midnight — which is
+ * still the *previous* day for anyone west of UTC. Parsing it as a local
+ * calendar day keeps the day the user actually clicked.
+ */
+export function parsePickerDate(value: Date | string): Date | null {
+  const parsed = value instanceof Date ? DateTime.fromJSDate(value) : DateTime.fromISO(value);
+  return parsed.isValid ? parsed.toJSDate() : null;
+}
+
+/**
  * Inverse of {@link photoTimestampToPickerDate}: turn a picker `Date` back into
  * an `exif_timestamp`, keeping the wall clock the user typed rather than
  * converting it to a real UTC instant.

--- a/apps/frontend/src/util/dateUtils.ts
+++ b/apps/frontend/src/util/dateUtils.ts
@@ -1,0 +1,68 @@
+import { DateTime } from "luxon";
+
+/**
+ * Helpers for reading and writing `exif_timestamp`.
+ *
+ * A photo's `exif_timestamp` is *local wall-clock time* — the time the camera
+ * showed when the shot was taken — which the backend deliberately stores tagged
+ * as UTC (`dt.replace(tzinfo=pytz.utc)` in `api/date_time_extractor.py`). A
+ * photo taken at 14:00 in Tokyo and one taken at 14:00 in Berlin both come back
+ * as `...T14:00:00Z`, because the timestamp says "14:00 where the camera was",
+ * not "this instant in time".
+ *
+ * Parsing that with `DateTime.fromISO(ts)` or `new Date(ts)` re-interprets it as
+ * a real instant and renders it in the *viewer's* zone, shifting every photo by
+ * the viewer's own UTC offset. Reading it back in UTC returns the wall clock
+ * unchanged, for every viewer, which is what the date rules promise.
+ */
+
+/** Parse an `exif_timestamp` into the wall-clock time the camera recorded. */
+export function parsePhotoTimestamp(timestamp: string): DateTime {
+  return DateTime.fromISO(timestamp, { zone: "utc" });
+}
+
+/**
+ * Convert an `exif_timestamp` into a JS `Date` whose *local* fields hold the
+ * wall clock, for date/time pickers that read `getHours()` and friends.
+ *
+ * The fields are copied across rather than converted, so the result is the same
+ * whatever zone the browser is in.
+ */
+export function photoTimestampToPickerDate(timestamp: string): Date | null {
+  const parsed = parsePhotoTimestamp(timestamp);
+  if (!parsed.isValid) {
+    return null;
+  }
+  return new Date(
+    parsed.year,
+    parsed.month - 1,
+    parsed.day,
+    parsed.hour,
+    parsed.minute,
+    parsed.second,
+    parsed.millisecond
+  );
+}
+
+/**
+ * Inverse of {@link photoTimestampToPickerDate}: turn a picker `Date` back into
+ * an `exif_timestamp`, keeping the wall clock the user typed rather than
+ * converting it to a real UTC instant.
+ */
+export function pickerDateToPhotoTimestamp(date: Date): string | null {
+  if (Number.isNaN(date.getTime())) {
+    return null;
+  }
+  return DateTime.fromObject(
+    {
+      year: date.getFullYear(),
+      month: date.getMonth() + 1,
+      day: date.getDate(),
+      hour: date.getHours(),
+      minute: date.getMinutes(),
+      second: date.getSeconds(),
+      millisecond: date.getMilliseconds(),
+    },
+    { zone: "utc" }
+  ).toISO();
+}

--- a/deploy/compose/docker-compose.yml
+++ b/deploy/compose/docker-compose.yml
@@ -65,6 +65,8 @@ services:
       - DB_PORT=5432
       - MAPBOX_API_KEY=${mapApiKey:-}
       - WEB_CONCURRENCY=${gunniWorkers:-1}
+      - WORKER_CONCURRENCY=${workerConcurrency:-}
+      - GUNICORN_TIMEOUT=${gunicornTimeout:-30}
       - SKIP_PATTERNS=${skipPatterns:-}
       - ALLOW_UPLOAD=${allowUpload:-false}
       - CSRF_TRUSTED_ORIGINS=${csrfTrustedOrigins:-}

--- a/deploy/compose/librephotos.env
+++ b/deploy/compose/librephotos.env
@@ -27,6 +27,20 @@ tag=latest
 # You’ll want to vary this a bit to find the best for your particular workload. Default is 2.
 gunniWorkers=2
 
+# Number of background workers that scan your library: thumbnails, face detection,
+# captioning. This is where nearly all of LibrePhotos' CPU and RAM goes.
+# Leave blank to use one worker per CPU core. Set a smaller number to leave
+# resources for other applications on the machine - this is the setting to reach
+# for, rather than a `cpus:` limit in docker-compose.yml, which only starves the
+# same number of workers. See docs.librephotos.com -> Limiting CPU and memory.
+workerConcurrency=
+
+# Seconds a single API request may take before gunicorn kills the worker. Raise
+# this if you cap the container's CPU and see "Worker was sent SIGKILL! Perhaps
+# out of memory?" in the backend log - on a throttled host that message usually
+# means the request was too slow, not that memory ran out. Default is 30.
+gunicornTimeout=30
+
 # You can set the database name. Did you know Libre Photos was forked from OwnPhotos?
 dbName=librephotos
 

--- a/deploy/docker/backend-gpu/entrypoint.sh
+++ b/deploy/docker/backend-gpu/entrypoint.sh
@@ -28,10 +28,16 @@ echo "Running backend server..."
 
 python manage.py qcluster 2>&1 | tee /logs/qcluster.log &
 
+# A request that outruns this is killed with SIGKILL and gunicorn's generic
+# "Perhaps out of memory?" message, which is misleading: on a CPU-capped host
+# the worker is usually just starved, not out of memory. Raise it if you limit
+# the container's CPU. Gunicorn's own default is 30.
+GUNICORN_TIMEOUT="${GUNICORN_TIMEOUT:-30}"
+
 if [[ "$DEBUG" = 1 ]]; then
     echo "development backend starting"
-    gunicorn --worker-class=gevent --max-requests 50 --reload --bind 0.0.0.0:8001 --log-level=info librephotos.wsgi 2>&1 | tee /logs/gunicorn_django.log
+    gunicorn --worker-class=gevent --max-requests 50 --timeout "$GUNICORN_TIMEOUT" --reload --bind 0.0.0.0:8001 --log-level=info librephotos.wsgi 2>&1 | tee /logs/gunicorn_django.log
 else
     echo "production backend starting"
-    gunicorn --worker-class=gevent --max-requests 50 --bind 0.0.0.0:8001 --log-level=info librephotos.wsgi 2>&1 | tee /logs/gunicorn_django.log
+    gunicorn --worker-class=gevent --max-requests 50 --timeout "$GUNICORN_TIMEOUT" --bind 0.0.0.0:8001 --log-level=info librephotos.wsgi 2>&1 | tee /logs/gunicorn_django.log
 fi

--- a/deploy/docker/backend/entrypoint.sh
+++ b/deploy/docker/backend/entrypoint.sh
@@ -27,10 +27,16 @@ echo "Running backend server..."
 
 python manage.py qcluster 2>&1 | tee /logs/qcluster.log &
 
+# A request that outruns this is killed with SIGKILL and gunicorn's generic
+# "Perhaps out of memory?" message, which is misleading: on a CPU-capped host
+# the worker is usually just starved, not out of memory. Raise it if you limit
+# the container's CPU. Gunicorn's own default is 30.
+GUNICORN_TIMEOUT="${GUNICORN_TIMEOUT:-30}"
+
 if [[ "$DEBUG" = 1 ]]; then
     echo "development backend starting"
-    gunicorn --worker-class=gevent --max-requests 50 --reload --bind 0.0.0.0:8001 --log-level=info librephotos.wsgi 2>&1 | tee /logs/gunicorn_django.log
+    gunicorn --worker-class=gevent --max-requests 50 --timeout "$GUNICORN_TIMEOUT" --reload --bind 0.0.0.0:8001 --log-level=info librephotos.wsgi 2>&1 | tee /logs/gunicorn_django.log
 else
     echo "production backend starting"
-    gunicorn --worker-class=gevent --max-requests 50 --bind 0.0.0.0:8001 --log-level=info librephotos.wsgi 2>&1 | tee /logs/gunicorn_django.log
+    gunicorn --worker-class=gevent --max-requests 50 --timeout "$GUNICORN_TIMEOUT" --bind 0.0.0.0:8001 --log-level=info librephotos.wsgi 2>&1 | tee /logs/gunicorn_django.log
 fi

